File: go/allergies.go
package allergies

var items = map[string]uint{
	"eggs":         1,
	"peanuts":      2,
	"shellfish":    4,
	"strawberries": 8,
	"tomatoes":     16,
	"chocolate":    32,
	"pollen":       64,
	"cats":         128,
}

func Allergies(allergies uint) []string {
	arr := []string{}
	for k := range items {
		if AllergicTo(allergies, k) {
			arr = append(arr, k)
		}
	}
	return arr
}

func AllergicTo(allergies uint, allergen string) bool {
	return allergies&items[allergen] > 0
}

File: ocaml/allergies.ml
type allergen =
  | Eggs
  | Peanuts
  | Shellfish
  | Strawberries
  | Tomatoes
  | Chocolate
  | Pollen
  | Cats

let allergenScores =
  [
    (Eggs, 1);
    (Peanuts, 2);
    (Shellfish, 4);
    (Strawberries, 8);
    (Tomatoes, 16);
    (Chocolate, 32);
    (Pollen, 64);
    (Cats, 128);
  ]

let allergic_to n a = List.assoc a allergenScores |> ( land ) n |> ( <> ) 0

let allergies n =
  List.filter_map
    (fun (a, _) -> if allergic_to n a then Some a else None)
    allergenScores
